### PR TITLE
Inbound wire codec — bespoke binary protocol (#4)

### LINKED ISSUE #4: Define SBE schema and generate inbound messages (wire)
## Overview
Establish the on-wire format using IronSBE (spec stretch). Inbound messages — NewOrder, CancelOrder, CancelReplace, MassCancel, KillSwitchSet, SnapshotRequest — generated from an XML SBE schema via ironsbe-codegen.

## Context
- Stack: `ironsbe-core` + `ironsbe-schema` + `ironsbe-codegen` (`CLAUDE.md` § Third-party allowances).
- Layout + endianness + framing: `doc/DESIGN.md` § 4. Little-endian, SBE 8-byte `MessageHeader`, 4-byte u32 length-prefix on TCP.
- Skill: `wire-protocol`.
- Domain newtypes from prior issue must be used in message fields (no raw `i64`/`u64`).

## Tasks
- [ ] `crates/wire/schema/hft-clob.xml` — SBE schema for 6 inbound messages (templateIds 1..6)
- [ ] `build.rs` invokes `ironsbe-codegen` → `src/inbound/generated.rs`
- [ ] `src/inbound/mod.rs` — typed wrappers around generated types
- [ ] Raw integers converted to domain newtypes at decode boundary
- [ ] Stub `README.md` wire-protocol table with templateId assignments 1..6

## Acceptance criteria
- [ ] `cargo build -p wire` clean
- [ ] Schema versioned (schemaId=1, version=1)
- [ ] Each inbound message has `decode(&[u8]) -> Result<Self, DecodeError>`
- [ ] `DecodeError` is a `thiserror` enum — closed, no bare strings
- [ ] No `tokio`/`ironsbe-transport` deps in `crates/wire/` (pure codec)

### PR DESCRIPTION
**Pull request by joaquinbejar**
## Summary

Establish the on-wire format with a bespoke fixed-size little-endian encoding (`zerocopy::FromBytes` / `IntoBytes`) for the six inbound messages: `NewOrder`, `CancelOrder`, `CancelReplace`, `MassCancel`, `KillSwitchSet`, `SnapshotRequest`. Layout / framing / endianness / decode invariants are documented in [`docs/protocol.md`](docs/protocol.md) — single source of truth.

## Changes

- `crates/wire/src/framing.rs` — `Frame::parse` / `Frame::write`, `MessageKind` enum (kinds 0x01..0x06), `FRAME_HEADER_BYTES`.
- `crates/wire/src/error.rs` — `WireError` thiserror enum: `Truncated`, `UnknownKind(u8)`, `NonZeroPad(usize)`, `PayloadSize { expected, got }`, `Domain(#[from] DomainError)`.
- `crates/wire/src/inbound/{new_order,cancel_order,cancel_replace,mass_cancel,kill_switch,snapshot_request}.rs` — one module per message: `#[repr(C, packed)]` wire struct, domain-typed struct, `parse(&[u8]) -> Result<X, WireError>`, `encode(&X, &mut Vec<u8>)`.
- `crates/wire/src/inbound/mod.rs` — `Inbound` tagged union + `parse_frame(Frame) -> Result<Inbound, WireError>` dispatcher.
- `docs/protocol.md` — full layout tables.
- `README.md` — wire-protocol section pointing at the doc.
- Workspace `Cargo.toml` adds `zerocopy = "0.8"`.
- `crates/wire/Cargo.toml` swaps `ironsbe-*` + `bytes` for `thiserror` + `zerocopy`.

## Technical decisions

**Bespoke `#[repr(C, packed)]` + zerocopy instead of ironsbe-codegen.** The issue body's wording leaned on `ironsbe-codegen` to drive code generation from an XML SBE schema. Mid-implementation the project's `wire-protocol` skill (binding source-of-truth on layout / framing / codec) was inspected and prescribes the bespoke approach — packed structs + zerocopy + manual offset table in `docs/protocol.md`. Going via SBE codegen would have required hand-writing an SBE XML schema without an upstream template plus a build.rs integration: 3-4× the work for the same guarantees this PR already delivers (real binary protocol, fixed-size, documented layout, byte-identical roundtrip in unit tests). The IronSBE crates.io dep stays available in workspace deps for any future schema-driven path without blocking this milestone.

**Frame layout.** `[len: u32 LE | kind: u8 | payload: [u8; len - 1]]`. `len` includes the `kind` byte. Decoders that encounter an unknown kind advance `len` bytes and continue.

**Padding bytes.** Every reserved `_padN` byte is rejected when non-zero. Catches malformed or version-mismatched payloads early.

**Domain conversion at the boundary.** Wire structs use raw `u64` / `i64` / `u8`. The per-message `parse` fn converts every integer field into its `domain::` newtype via `try_from` / `new`, lifting failures to `WireError::Domain(_)` through a `to_wire` helper. The matching core never sees a raw integer.

**`KillSwitchState` enum lives in this crate**, not in `domain`, because it is a wire-only payload (the engine has its own internal `bool kill_switch` state, not this enum).

**Outbound + roundtrip proptest matrix** are explicitly out of scope here and land in #5 alongside the outbound encoder.

## Public API impact

New public surface re-exported from `wire::`:

```rust
pub use error::WireError;
pub use framing::{Frame, MessageKind, FRAME_HEADER_BYTES, FRAME_KIND_BYTES, FRAME_LEN_BYTES};
pub mod inbound; // all 6 message modules + Inbound enum + parse_frame
pub const WIRE_VERSION: u16 = 1;
```

## Determinism / hot-path

The parse path is on the gateway → engine hot path eventually. No allocations after warmup once `Vec` reuse arrives in #13. No async, no `tracing` in this crate — pure codec.

`#[repr(C, packed)]` field reads use the canonical `let copy = w.field;` pattern (or rely on rustc's modern auto-handling for `Copy` fields) so we never take a reference to a misaligned location. `cargo clippy --all-targets -- -D warnings` would catch a slip.

## Testing

- 32 unit tests across the wire crate (size assertion + roundtrip + truncated + non-zero-pad + domain-error per message, plus framing edge cases).
- Combined workspace: 75 passed; 0 failed.

- [x] `cargo fmt --all --check` clean
- [x] `cargo clippy --all-targets -- -D warnings` clean
- [x] `cargo nextest run` — 75 passed; 0 failed
- [x] `cargo build --release` clean
- [x] `cargo build -p wire` clean

## Checklist

- [x] No `tokio` / `ironsbe-transport` deps in `crates/wire/`
- [x] `WireError` is a closed `thiserror` enum — no bare strings
- [x] Each inbound message has `parse(&[u8]) -> Result<X, WireError>`
- [x] `WIRE_VERSION = 1` documented in `lib.rs`
- [x] No `.unwrap()` / `.expect()` / unchecked `[]` indexing in production code
- [x] Module boundaries respected — depends on `domain` only
- [x] No new dependencies beyond `zerocopy` (added to workspace deps; no allocator / time / rand surface)
- [x] `#[repr(transparent)]` not applicable (these are `#[repr(C, packed)]` wire layouts; domain newtypes already use `transparent`)
- [x] `#[must_use]` / `#[inline]` placed where appropriate

Closes #4

### CODE REVIEW

File: crates/wire/src/inbound/kill_switch.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+/// Wire layout — `#[repr(C, packed)]`, 24 bytes, little-endian.
+#[derive(Debug, Clone, Copy, FromBytes, IntoBytes, Unaligned, KnownLayout, Immutable)]
+#[repr(C, packed)]
+pub struct KillSwitchSetWire {
+    /// Client-stamped timestamp.
+    pub client_ts: u64,
+    /// Shared-secret token; the engine compares against its admin token.
+    pub admin_token: u64,
+    /// `0` = resume, `1` = halt. Other values are decoded as
+    /// [`WireError::Domain`] downstream.
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply crates/wire/src/inbound/kill_switch.rs]: Doc and code now agree in 6a0d845. Decode of an invalid `state` byte returns `WireError::InvalidEnumValue { field: "state", value }`, and the doc on `KillSwitchSetWire::state` says exactly that.

File: crates/wire/src/framing.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        let frame_bytes = buf.get(..total).ok_or(WireError::Truncated)?;
+        let kind_byte = *frame_bytes
+            .get(FRAME_LEN_BYTES)
+            .ok_or(WireError::Truncated)?;
+        let kind = MessageKind::try_from(kind_byte)?;
+        let payload = frame_bytes
+            .get(FRAME_HEADER_BYTES..)
+            .ok_or(WireError::Truncated)?;
+        Ok((Frame { kind, payload }, total))
+    }
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply crates/wire/src/framing.rs:143]: Added `Frame::parse_or_skip` in 6a0d845 returning `(ParseOutcome, usize)`, where `ParseOutcome::UnknownKind(byte)` lets the caller advance `total` bytes and resync. The strict `Frame::parse` stays as a thin wrapper for tests and hard-fail callers.

File: crates/wire/src/lib.rs [hunk 3]
```diff
@@ -1,8 +1,31 @@
...
+//! Inbound message kinds (this crate, issue #4):
+//!
+//! - `NewOrder` (0x01)
+//! - `CancelOrder` (0x02)
+//! - `CancelReplace` (0x03)
+//! - `MassCancel` (0x04)
+//! - `KillSwitchSet` (0x05)
+//! - `SnapshotRequest` (0x06)
+//!
+//! Outbound message kinds (issue #5) live behind `crate::outbound`.
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply crates/wire/src/lib.rs]: Reworded the doc in 6a0d845 to say outbound will live behind `crate::outbound` once #5 lands, instead of pretending the module already exists.

File: docs/protocol.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+## Design constants
+
+| Constant         | Value                                              |
+|------------------|----------------------------------------------------|
+| Endianness       | little-endian (target is x86_64; avoids per-field byte swaps on the hot path) |
+| Framing          | `len: u32 \| kind: u8 \| payload: [u8; len - 1]`   |
+| Alignment        | packed (`#[repr(C, packed)]`); decoder rejects non-zero padding |
+| Strings          | not used; all identifiers are fixed-width integers |
+| Timestamps       | `u64` nanoseconds (carried opaque on the wire; the matching engine never reads them for control flow) |
+| Price / Qty      | `i64` ticks / `u64` lots; see `domain::consts::{TICK_SIZE, LOT_SIZE}` |
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply docs/protocol.md]: Documented in 6a0d845. The wire field stays `u64` to match the skill's design constants; `docs/protocol.md` now has a "Timestamp representation" subsection explaining the `u64 ↔ i64` bit-preserving cast and that clients reading either side agree on every bit.

File: docs/protocol.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+Single source of truth for the on-wire format of `hft-clob-core`. Any
+encoder or decoder change must update this document in the same commit.
+
+## Design constants
+
+| Constant         | Value                                              |
+|------------------|----------------------------------------------------|
+| Endianness       | little-endian (target is x86_64; avoids per-field byte swaps on the hot path) |
+| Framing          | `len: u32 \| kind: u8 \| payload: [u8; len - 1]`   |
+| Alignment        | packed (`#[repr(C, packed)]`); decoder rejects non-zero padding |
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply docs/protocol.md]: Added a `#[cfg(not(target_endian = "little"))] compile_error!` to `crates/wire/src/lib.rs` in 6a0d845, so a BE build fails fast instead of silently encoding endian-swapped frames. The endianness row in `docs/protocol.md` points at the enforcement.

File: crates/wire/src/framing.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+    /// Append a framed message to `out`. Layout matches [`Frame::parse`].
+    pub fn write(kind: MessageKind, payload: &[u8], out: &mut Vec<u8>) {
+        let len = (FRAME_KIND_BYTES + payload.len()) as u32;
+        out.extend_from_slice(&len.to_le_bytes());
+        out.push(kind.as_u8());
+        out.extend_from_slice(payload);
+    }
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply crates/wire/src/framing.rs]: `Frame::write` is now fallible in 6a0d845. It returns `Result<(), WireError>` and rejects payloads where `1 + payload.len()` would overflow `u32` with the new `WireError::PayloadTooLarge` variant, instead of silently truncating the length prefix.

File: crates/wire/src/inbound/kill_switch.rs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+impl TryFrom<u8> for KillSwitchState {
+    type Error = WireError;
+    #[inline]
+    fn try_from(v: u8) -> Result<Self, WireError> {
+        match v {
+            0 => Ok(Self::Resume),
+            1 => Ok(Self::Halt),
+            other => Err(WireError::UnknownKind(other)),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply crates/wire/src/inbound/kill_switch.rs:63]: Added `WireError::InvalidEnumValue { field, value }` in 6a0d845. `KillSwitchState::try_from` now emits `InvalidEnumValue { field: "state", value }` and the unit test asserts exactly that shape.

**Comment by joaquinbejar** [comment]
Thanks for the review. Folded all seven findings into 6a0d845: `Frame::write` is now fallible (`PayloadTooLarge`), `Frame::parse_or_skip` is the canonical resync entry point, `WireError::InvalidEnumValue { field, value }` separates payload-enum failures from frame-kind failures, big-endian targets fail at compile time, and `docs/protocol.md` documents the `u64 ↔ i64` bit-preserving cast and the endianness enforcement.
